% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/radiator_dots.R
\name{radiator_dots}
\alias{radiator_dots}
\title{radiator_dots}
\usage{
radiator_dots(
  func.name = as.list(sys.call())[[1]],
  fd = NULL,
  args.list = NULL,
  dotslist = NULL,
  keepers = c("subsample.markers.stats", "force.stats", "id.stats", "subsample",
    "filter.reproducibility", "filter.individuals.missing",
    "filter.individuals.heterozygosity", "filter.individuals.coverage.total",
    "filter.individuals.coverage.median", "filter.individuals.coverage.iqr",
    "filter.common.markers", "filter.monomorphic", "filter.mac", "filter.coverage", "dp",
    "filter.genotyping", "filter.snp.position.read", "filter.snp.number",
    "filter.short.ld", "filter.long.ld", "long.ld.missing", "ld.method", "ld.figures",   
      "detect.mixed.genomes", "ind.heterozygosity.threshold", "detect.duplicate.genomes",
    "filter.hwe", "filter.strands", "random.seed", "path.folder", "filename",
    "parameters", "blacklist.genotypes", "erase.genotypes", "gt", "gt.bin", "gt.vcf",
    "gt.vcf.nuc", "pop.levels", "pop.labels", "pop.select", "blacklist.id",
    "markers.info", "keep.allele.names", "keep.gds", "calibrate.alleles", "vcf.metadata",
    "vcf.stats", "wide", "whitelist.markers", "write.tidy", "missing.memory",
    "dart.sequence", "internal",      "heatmap.fst", "tidy.check", "tidy.vcf",
    "tidy.dart"),
  deprecated = c("maf.thresholds", "common.markers", "max.marker", "monomorphic.out",
    "snp.ld", "filter.call.rate", "filter.markers.coverage", "filter.markers.missing",
    "number.snp.reads", "mixed.genomes.analysis", "duplicate.genomes.analysis",
    "ref.calibration"),
  verbose = TRUE
)
}
\arguments{
\item{func.name}{Default: \code{as.list(sys.call())[[1]]}.}

\item{fd}{(optional) Default: \code{rlang::fn_fmls_names()}.}

\item{args.list}{(optional) Default:\code{args.list = as.list(environment())}.}

\item{dotslist}{The dots dots dots arguments captured.
Default: \code{
dotslist = rlang::dots_list(..., .homonyms = "error", .check_assign = TRUE)}.}

\item{keepers}{(optional) The arguments that are used inside the function.
Default: \code{keepers = c(
"subsample.markers.stats", "subsample",
"filter.reproducibility", "filter.individuals.missing",
"filter.individuals.heterozygosity", "filter.individuals.coverage.total",
"filter.individuals.coverage.median", "filter.individuals.coverage.iqr",
"filter.common.markers", "filter.monomorphic", "filter.mac",
"filter.coverage", "filter.genotyping", "filter.snp.position.read",
"filter.snp.number", "filter.short.ld", "filter.long.ld", "long.ld.missing",
"ld.method", "detect.mixed.genomes", "ind.heterozygosity.threshold",
"detect.duplicate.genomes",
"filter.hwe", "filter.strands", "random.seed", "path.folder", "filename",
"blacklist.genotypes", "erase.genotypes",
"gt", "gt.bin", "gt.vcf", "gt.vcf.nuc",
"pop.levels", "pop.labels", "pop.select", "blacklist.id",
"markers.info", "keep.allele.names", "keep.gds", "calibrate.alleles",
"vcf.metadata", "vcf.stats", "id.stats", "dp",
"whitelist.markers",
"write.tidy",
"dart.sequence",
"missing.memory",
"internal", "heatmap.fst", "tidy.check", "tidy.vcf", "tidy.dart")}.}

\item{deprecated}{(optional) radiator's deprecated arguments.
Default: \code{deprecated = c("maf.thresholds", "common.markers",
"max.marker","monomorphic.out", "snp.ld", "filter.call.rate",
"filter.markers.coverage", "filter.markers.missing", "number.snp.reads",
"mixed.genomes.analysis", "duplicate.genomes.analysis", "maf.data", "ref.calibration")}.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE}
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}
}
\description{
Extract and assign the dots-dots-dots
}
