% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/snprelate.R
\name{write_snprelate}
\alias{write_snprelate}
\title{Write a SNPRelate object from a tidy data frame}
\usage{
write_snprelate(data, biallelic = TRUE, filename = NULL, verbose = TRUE)
}
\arguments{
\item{data}{A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
\emph{How to get a tidy data frame ?}
Look into \pkg{radiator} \code{\link{tidy_genomic_data}}.
\strong{The genotypes are biallelic.}}

\item{biallelic}{(logical, optional) If you already know that the data is
biallelic use this argument to speed up the function.
Default: \code{biallelic = TRUE}.}

\item{filename}{(optional) The file name of the Genomic Data Structure (GDS) file.
radiator will append \code{.gds} to the filename.
If filename chosen is already present in the
working directory, the default \code{radiator_snprelate_datetime.gds} is chosen.
Default: \code{filename = NULL}.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE}
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}
}
\value{
An object in the global environment of class
\code{"SNPGDSFileClass", "gds.class"} and
a file in the working directory.
}
\description{
Write a \href{https://github.com/zhengxwen/SNPRelate}{SNPRelate}
object from a tidy data frame.
Used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
and might be of interest for users.
}
\examples{
\dontrun{
require(SNPRelate)
data.gds <- radiator::write_snprelate(data = "shark.rad")
}
}
\references{
Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS.
A high-performance computing toolset for relatedness and principal component
analysis of SNP data. Bioinformatics. 2012;28: 3326-3328.
doi:10.1093/bioinformatics/bts606
}
\seealso{
\href{https://github.com/zhengxwen/SNPRelate}{SNPRelate}
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
