<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Write a COLONY input file — write_colony • radiator</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Write a COLONY input file — write_colony" />
<meta property="og:description" content="Write a COLONY input file." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Write a <code>COLONY</code> input file</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/write_colony.R'><code>R/write_colony.R</code></a></small>
    <div class="hidden name"><code>write_colony.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Write a <code>COLONY</code> input file.</p>
    </div>

    <pre class="usage"><span class='fu'>write_colony</span><span class='op'>(</span>
  <span class='va'>data</span>,
  strata <span class='op'>=</span> <span class='cn'>NULL</span>,
  sample.markers <span class='op'>=</span> <span class='cn'>NULL</span>,
  pop.select <span class='op'>=</span> <span class='cn'>NULL</span>,
  allele.freq <span class='op'>=</span> <span class='cn'>NULL</span>,
  inbreeding <span class='op'>=</span> <span class='fl'>0</span>,
  mating.sys.males <span class='op'>=</span> <span class='fl'>0</span>,
  mating.sys.females <span class='op'>=</span> <span class='fl'>0</span>,
  clone <span class='op'>=</span> <span class='fl'>0</span>,
  run.length <span class='op'>=</span> <span class='fl'>2</span>,
  analysis <span class='op'>=</span> <span class='fl'>1</span>,
  allelic.dropout <span class='op'>=</span> <span class='fl'>0</span>,
  error.rate <span class='op'>=</span> <span class='fl'>0.02</span>,
  print.all.colony.opt <span class='op'>=</span> <span class='cn'>FALSE</span>,
  random.seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>,
  parallel.core <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>-</span> <span class='fl'>1</span>,
  filename <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>(4 options) A file or object generated by radiator:</p><ul>
<li><p>tidy data</p></li>
<li><p>Genomic Data Structure (GDS)</p></li>
</ul>

<p><em>How to get GDS and tidy data ?</em>
Look into <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>,
<code><a href='read_vcf.html'>read_vcf</a></code> or
<code><a href='tidy_vcf.html'>tidy_vcf</a></code>.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(path or object) The strata file or object.
Additional documentation is available in <code><a href='read_strata.html'>read_strata</a></code>.
Use that function to whitelist/blacklist populations/individuals.
Option to set <code>pop.levels/pop.labels</code> is also available.</p></td>
    </tr>
    <tr>
      <th>sample.markers</th>
      <td><p>(number) <code>COLONY</code> can take a long time to run,
use a random subsample of your markers to speed test <code>COLONY</code>
e.g. <code>sample.markers = 500</code> to use only 500 randomly chosen markers.
Default: <code>sample.markers = NULL</code>, will use all markers.</p></td>
    </tr>
    <tr>
      <th>pop.select</th>
      <td><p>(optional, string) Selected list of populations for
the analysis. e.g. <code>pop.select = c("QUE", "ONT")</code> to select <code>QUE</code>
and <code>ONT</code> population samples (out of 20 pops).
Default: <code>pop.select = NULL</code></p></td>
    </tr>
    <tr>
      <th>allele.freq</th>
      <td><p>(optional, string) Allele frequency can be computed from
a select group.
e.g. <code>allele.freq = "QUE"</code> or <code>allele.freq = c("QUE", "ONT")</code>.
Using <code>allele.freq = "overall"</code> will use all the samples to compute the
allele frequency.
Default: <code>allele.freq = NULL</code>, will not compute allele frequency.</p></td>
    </tr>
    <tr>
      <th>inbreeding</th>
      <td><p>(boolean) 0/1 no inbreeding/inbreeding.
Default: <code>inbreeding = 0</code></p></td>
    </tr>
    <tr>
      <th>mating.sys.males</th>
      <td><p>(boolean) Mating system in males.
0/1 polygyny/monogyny.
Default: <code>mating.sys.males = 0</code>.</p></td>
    </tr>
    <tr>
      <th>mating.sys.females</th>
      <td><p>(boolean) Mating system in females.
0/1 polygyny/monogyny.
Default: <code>mating.sys.females = 0</code>.</p></td>
    </tr>
    <tr>
      <th>clone</th>
      <td><p>(boolean) Should clones and duplicated individuals be inferred.
0/1, yes/no. Default: <code>clone = 0</code>.</p></td>
    </tr>
    <tr>
      <th>run.length</th>
      <td><p>(integer) Length of run. 1 (short), 2 (medium), 3 (long),
4 (very long). Start with short or medium run and consider longer run if your
estimates probability are not stable or really good.
Default: <code>run.length = 2</code>.</p></td>
    </tr>
    <tr>
      <th>analysis</th>
      <td><p>(integer) Analysis method.
0 (Pairwise-Likelihood Score), 1 (Full Likelihood),
2 (combined Pairwise-Likelihood Score and Full Likelihood).
Default: <code>analysis = 1</code>.</p></td>
    </tr>
    <tr>
      <th>allelic.dropout</th>
      <td><p>Locus allelic dropout rate.
Default : <code>allelic.dropout = 0</code>.</p></td>
    </tr>
    <tr>
      <th>error.rate</th>
      <td><p>Locus error rate.
Default:<code>error.rate = 0.02</code>.</p></td>
    </tr>
    <tr>
      <th>print.all.colony.opt</th>
      <td><p>(logical) Should all <code>COLONY</code> options be printed in the file.</p>
<p><strong>This require manual curation, for the file to work directly with <code>COLONY</code></strong>.
Default = <code>print.all.colony.opt = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>random.seed</th>
      <td><p>(integer, optional) For reproducibility, set an integer
that will be used inside the function that requires randomness. With default,
a random number is generated and printed in the appropriate output.
Default: <code>random.seed = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(optional, logical) When <code>verbose = TRUE</code>
the function is a little more chatty during execution.
Default: <code>verbose = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>Name of the acronym for filenaming in the working directory.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>COLONY</code> file in your working directory (2 if you selected imputations arguments...)</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>It is highly recommended to read (twice!) the user guide distributed with
<code>COLONY</code> to find out the details for input and output of the software.</strong></p>
<p>Not all options are provided here.</p>
<p>But to ease the process, all the required options to properly run <code>COLONY</code>
will be printed in the file written in your working directory.
Change the values accordingly and wisely.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Jones OR, Wang J (2010) COLONY: a program for parentage and
sibship inference from multilocus genotype data.
Molecular Ecology Resources, 10, 551–555.</p>
<p>Wang J (2012) Computationally Efficient Sibship and
Parentage Assignment from Multilocus Marker Data. Genetics, 191, 183–194.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>COLONY</code> is available on Jinliang Wang web site
<a href='https://www.zsl.org/science/software/colony'>https://www.zsl.org/science/software/colony</a></p>

<p><a href='https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html#colony'>colony installation instructions</a></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># Simplest way to run the function with a tidy dataset:</span>
<span class='va'>colony.file</span> <span class='op'>&lt;-</span> <span class='fu'>radiator</span><span class='fu'>::</span><span class='fu'>write_colony</span><span class='op'>(</span>data <span class='op'>=</span> <span class='st'>"turtle.data.rad"</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


